lexibank_chaconcolumbian.py

```diff
@@ -17,45 +17,105 @@ class ChaconColConcept(Concept):
-
-            for concept in self.conceptlist.concepts.values():
-                ds.add_concept(
-                    ID=concept.id,
-                    Name=concept.english,
-                    Concepticon_ID=concept.concepticon_id,
-                    Concepticon_Gloss=concept.concepticon_gloss,
-                    Spanish=concept.attributes["spanish"],
-                    Gloss_in_source=concept.attributes["gloss_in_source"],
-                    Category=concept.attributes["category"],
+    concept_class = CustomConcept
+
+    def cmd_makecldf(self, args):
+        # Read the raw data
+        wl_filename = self.raw_dir / "Huber_filtered_130_basic_cult_voc"
+        wl = lingpy.Wordlist(wl_filename.as_posix())
+
+        # Write sources
+        args.writer.add_sources()
+
+        # Write languages
+        languages = args.writer.add_languages(lookup_factory="Name")
+
+        # Write concepts
+        concepts = {}
+        for concept in self.conceptlists[0].concepts.values():
+            concept_cldf_id = "%s_%s" % (
+                concept.id.split("-")[-1],
+                slug(concept.english),
+            )
+            args.writer.add_concept(
+                ID=concept_cldf_id,
+                Name=concept.english,
+                Concepticon_ID=concept.concepticon_id,
+                Concepticon_Gloss=concept.concepticon_gloss,
+                Spanish=concept.attributes["spanish"],
+                Gloss_in_source=concept.attributes["gloss_in_source"],
+                Category=concept.attributes["category"],
+            )
+            concepts[concept.attributes["gloss_in_source"]] = concept_cldf_id
+
+        # Hard-coded fixes to segment errors in raw source
+        segments = {
+            " #h ": " h ",
+            " #s ": " s ",
+            " aⁱ ": " a j ",
+            " aᵘ ": " a w ",
+            " eⁱ ": " e j ",
+            " hh ": " h h ",
+            " hs ": " h s ",
+            " i̵ ": " ɨ ",
+            " ĩ̵ ": " ɨ̃ ",
+            " ĩ̵: ": " ɨ̃ɨ̃ ",
+            " i̵: ": " ɨɨ ",
+            " ĩ̵ĩ̵ ": " ɨ̃ɨ̃ ",
+            " i̵i̵ ": " ɨɨ ",
+            " i̵i̵i̵ ": " ɨ ɨɨ ",
+            " ɩ ": " ɪ ",
+            " ɩɩ ": " ɪɪ ",
+            " k jh ": " kʲ h ",
+            " ll ": " l l ",
+            " ls ": " l s ",
+            " mm ": " m m ",
+            " nn ": " n n  ",
+            " nss ": " n s ",
+            " ŋŋ ": " ŋ ŋ ",
+            " oᵘ ": " o w ",
+            " rh ": " r h ",
+            " rr ": " r r ",
+            " rs ": " r s ",
+            " ɾh ": " ɾ h ",
+            " ɾs ": " ɾ s ",
+            " sh ": " s h ",
+            " ss ": " s s ",
+            " vh ": " vʰ ",
+            " vhs ": " vʰ s ",
+            " xh ": " x h ",
+        }
+
+        # write lexemes
+        for entry in progressbar(wl, desc="makecldf"):
+            if wl[entry, "concept"]:
+                # fix errors in segments: as in some cases there is no
```

Comment:
I do not believe that list-comprehension would not work here, @tresoldi. This approach, even if it fixes the entries, is a bit too shaky. What you could do is: use this approach to produce an enhanced list of segments, and then replace your previous one.

Reply:
But how would you solve with non shaky code fixes such as `"ls"` to `"l s"` (one segment to two) or `"k jh"` to `"kʲ h"` (same number of segments, but `k` is replaced if and only if it is followed by `jh`)?

One alternative I considered was to map segments to lists (even if lists of a single element) and then flatten the list, either with itertools or in the same comprehension (i.e., `[grapheme for group in segments for grapheme in group]`), but opted for the simpler approach of just replacing.

Reply:
Ah okay, so the question is: how many of these problems are there in the end? If you add the alignments, it is clear that you cannot use your approach anyway. So how many of these cases are there in the data? If we talk about 100 examples, I would say, you have two possibilities:

1. put the form into `etc/lexemes.tsv`, but add a third column, called Segments, another one called Alignments, and you can lookup from `self.lexemes`
2. make the lookup as part of a dictionary inside the scripts for these terms.

In fact, since your code now already produces this extended `lexemes.tsv`, you could just run the code, filter all entries where it really DOES do something, and put your replacments into `lexemes.tsv`. You see: your approach is good in fixing the issues. But it is not good, as it lacks explicitness. And the same in fact also applies to what I did with `bodtkhobwa`.

Reply:
It is almost all the cases, with the exception of central high vowels and two more cases. In most cases it looks like an artifact from removal of tones.

I would much prefer having them in `lexemes.tsv`, I'll rewrite it.

Reply:
Did you check if the alignments work after all? If they are clearly NOT human-curated, I would say: drop alignments (we can later automat them). And in THIS case, you could even just use the segmented data you already have and add an orthography profile to segment the data another time. 

But please: do what costs you less time here. Both solutions we discussed seem pragmatic enough, but also explicit enough. But consider dropping the alignments, as this will mean less work, and it seems they are not worth too much pain here.

Reply:
There are no alignments, do you mean the segments? I'd guess they were manually corrected from the result of some automatic pre-processing.

Now that I noticed that `self.lexemes` does not pick up other columns ([here](https://github.com/lexibank/pylexibank/blob/434fa0890ba15e3f1a3ceb446edee0456b020bd2/src/pylexibank/dataset.py#L114)), I am not sure about the least time-expensive solution. The "perfect" one would be to write a profile based on the segmented data we now have.

A quicker alternative, with no changes to `pylexibank`, could be setting up an `etc/segments.csv` file, which I would need to load manually into `segments` (two-three lines) and would be used in `add_form_with_segments`:

```python
Segments=segments.get(tuple(wl[entry, "tokens"]), wl[entry, "segments"]),
```

Tuple conversion is needed because basictypes.list is non-hasheable. Would that be ok?

Reply:
There are alignments, I mean, your code adds alignments from the wordlist file.

Reply:
Oh, right, in the cognates! The alignments there are following the segmentation in the same wordlist. I'll remove them.

Reply:
Okay, this is important. I think that the major problems you identify in the data are not as problematic as you think.

The reason is: the "ls", for example, occurs in one instance, and this word is not cognate with other words. Same holds for "k jh", only one entry. And I am not sure that your solution is correct here, if you look at the cognate sets.

I recommend you to have a look at the basic data file through edictor, and check for the instances, using the search interface (filter columns), where you can write something like "tokens=jh" to see how many instances you find that contain this. You'll be surprised how much of what seems to be a problem can be resolved by a direct construct, e.g., writing "jh/j" instead of "jh", as we do not know what it  means here. 

The alignments are in fact good, I checked them in edictor now, they yield nice corrspondence patterns. The segments lookup procedure needs to be used both on the alignments and on the segments.

Reply:
One more thing: I checked teh alignmetns again, and we can drop them. 

Furthermore, all changes proposed by you in the segments lookup should be checked against the counterpart. the "hh" for example should be rendered as "h + h", as there's a morpheme boundary, etc.

Reply:
But this way I am back to the issue of unbalanced number of segments: "jh" to "jh/j" is patent, but would I do in a comprehension "hh" to "h + h"?

Reply:
No, if we drop the alignments, you can live with changing the segments, right?

Reply:
I just emphasize: most of the few changes you introduced there occur only one or two times, so they are not doing much to the data.

Reply:
Maybe should I just add to this PR and show what I mean?

Reply:
In fact, I'll do that now. The modifications also introduce BIPA-incompatibilities, so I'll update this now.

Reply:
To avoid confusion:

- We drop the alignments from the CLDF data
- I review my corrections by checking all instances in the source alignments, with Edictor
- I compile a list of form->segment replacements in `etc/segments.tsv`, so we don't need changes to `pylexibank`
- I load a `segment` lookup in code and do a `Segments=segments.get(form, wl_segments)` when adding

Sorry I misunderstood something.

Reply:
Ok, I'll wait.

Reply:
no direct misunderstandings, I just don't agree with your replacments in the segments, since I already checked most of the instances in edictor. I'll try to push my suggestions soon.

Reply:
Okay, @tresoldi, please check what I did. I removed the alignments, and I made all segmentation issues explicit. As there were many variants that are accepted by CLTS but not standard clts. The reason is that this data stems from the old version of Huber with bad segmentation.

Reply:
@tresoldi, if you rerun to make the CLDF (with the respective versions of concepticon, etc.) this would be nice, so you can push the updated CLDF, and close this issue PR, after merging.

Reply:
What is essential, as we can see from this example, is to check all segments in Transcription.md. In this case, this revealed far more problems than the tests of pylexibank would have told us.

Reply:
I'm on it.